codaco/Network-Canvas
# Ensure uids exist for all nodes in app

Resolves #415.

This standardizes the behavior of primary keys for nodes in the app:

- Each node imported from external data are given a unique identifier, unless it already has one
  + UIDs are stable across imports as long as input data is stable (UID == digest of contents)
  + This is done up front, at least for now (may need to change with #413)
- If external data defines PK values, they are assumed to be unique.
- A protocol author may define a different key for each data source, using `"nodes_primary_key"`. See the example of `studentId` in the dev protocol.
- The convention (currently; see definition in modules/network) is to use the `uid` prop as an identifier. Nodes always have this property defined internally, regardless of the external data's primary key.
- Search interface can now rely on UIDs for result display

### Things to test
- Add nodes from different sources, e.g.
  + `previousInterview` has existing `uid`s
  + `schoolPupils` uses the same IDs, but uses a custom PK ("studentId") 
  + `HIVServices` and `venue_list` do not define any primary keys
- In the dev protocol, delete UIDs from "previousInterview". The first node generator should behave well (e.g., not contain an infinite number of the same node as you drag nodes out).
- Changing the PK prop ("uid") used internally by the app shouldn't matter
- Reloading the app should not change results shown in Search or other node generators
- View generated data by downloading the export file at the end

Edit: this also fixes #605, as UIDs should override anything in additionalAttributes (which cannot be unique).
